server/main_server/src/main.cpp

```diff
@@ -0,0 +1,1342 @@
+            return;
+        }
+
+        _time_start = now;
+
+        const auto hostname = boost::asio::ip::host_name();
+        std::optional<std::string> leader;
+        std::optional<std::string> successor;
+
+        try {
+            irods::experimental::client_connection conn;
+            leader = get_delay_server_leader(conn);
+            successor = get_delay_server_successor(conn);
+
+            if (leader && successor) {
+                log_server::debug("{}: Delay server leader [{}] and successor [{}].", __func__, *leader, *successor);
+
+                // 4 cases:
+                // L  S
+                // ----
+                // 0  0 (invalid state / no-op)
+                // 1  0 (migration complete)
+                // 0  1 (leader has shut down ds, successor launching ds)
+                // 1  1 (leader running ds, admin requested migration)
+
+                // This server is the leader and may be running a delay server.
+                if (hostname == *leader) {
+                    if (hostname == *successor) {
+                        launch_delay_server(_write_to_stdout, _enable_test_mode);
+
+                        // Clear successor entry in catalog. This isn't necessary, but helps
+                        // keep the admin from becoming confused.
+                        if (g_pid_ds > 0) {
+                            set_delay_server_migration_info(conn, KW_DELAY_SERVER_MIGRATION_IGNORE, "");
+                        }
+                    }
+                    else if (!successor->empty()) {
+                        // Migration requested. Stop local delay server if running and clear the
+                        // leader entry in the catalog.
+                        if (g_pid_ds > 0) {
+                            kill(g_pid_ds, SIGTERM);
+
+                            // Wait for delay server to complete current tasks. Depending on the workload, this
+                            // could take minutes, hours, days to complete.
+                            int status = 0;
+                            waitpid(g_pid_ds, &status, 0);
+                            log_server::info("Delay server has completed shutdown [exit_code={}].", WEXITSTATUS(status));
+                            g_pid_ds = 0;
+                        }
+
+                        // TODO This isn't necessary if we can get task locking working (i.e. delay servers
+                        // lock tasks just before execution).
+                        //
+                        // TODO However, we can uncomment the following if the server can confirm the host
+                        // doesn't exist in the zone.
+                        //
+                        // Clear the leader entry. This acts as a signal to the successor server
+                        // that it is safe to launch the delay server.
+                        //set_delay_server_migration_info(conn, "", KW_DELAY_SERVER_MIGRATION_IGNORE);
```

Comment:
Handle TODOs.

Reply:
this discussion still needed?

Reply:
Yes.

We should look at the new implementation and make sure it's correct. The TODO comments are worth discussing too.

Reply:
Added checkbox to PR description.